src/deluge/gui/views/instrument_clip_view.cpp

```diff
@@ -1282,93 +1282,112 @@ const uint32_t auditionPadActionUIModes[] = {UI_MODE_AUDITIONING,
 
 ActionResult InstrumentClipView::padAction(int32_t x, int32_t y, int32_t velocity) {
 
-	if (x == 15 && y == 2 && velocity > 0
+if (x == 15 && y == 2 && velocity > 0
 	    && runtimeFeatureSettings.get(RuntimeFeatureSettingType::DrumRandomizer) == RuntimeFeatureStateToggle::On) {
 		int32_t numRandomized = 0;
-		for (int32_t i = 0; i < 8; i++) {
-			if (getCurrentUI() == this && this->auditionPadIsPressed[i]) {
-				if (getCurrentInstrumentType() != InstrumentType::KIT) {
-					continue;
-				}
-				AudioEngine::stopAnyPreviewing();
-				Drum* drum = getCurrentInstrumentClip()->getNoteRowOnScreen(i, currentSong)->drum;
-				if (!drum || drum->type != DrumType::SOUND) {
-					continue;
-				}
-				SoundDrum* soundDrum = (SoundDrum*)drum;
-				MultiRange* r = soundDrum->sources[0].getRange(0);
-				if (r == NULL || ((MultisampleRange*)r)->sampleHolder.audioFile == NULL) {
-					continue;
-				}
-				AudioFileHolder* afh = r->getAudioFileHolder();
-				if (afh == NULL) {
-					continue;
-				}
+		if (sdRoutineLock) {
+			return ActionResult::REMIND_ME_OUTSIDE_CARD_ROUTINE;
+		}
+		if (currentSong->currentClip->output->type != InstrumentType::KIT) {
+			return ActionResult::DEALT_WITH;
+		}
+		AudioEngine::stopAnyPreviewing();
 
-				static int32_t MaxFiles = 25;
-				String fnArray[MaxFiles];
-				char const* currentPathChars = afh->filePath.get();
-				if (currentPathChars == NULL) {
-					continue;
-				}
-				char const* slashAddress = strrchr(currentPathChars, '/');
-				if (slashAddress) {
-					int32_t slashPos = (uint32_t)slashAddress - (uint32_t)currentPathChars;
-					String dir;
-					dir.set(&afh->filePath);
-					dir.shorten(slashPos);
-					FRESULT result = f_opendir(&staticDIR, dir.get());
-					FilePointer thisFilePointer;
-					int32_t numSamples = 0;
-
-					if (result != FR_OK) {
-						display->displayError(ERROR_SD_CARD);
-						return ActionResult::DEALT_WITH;
-					}
-					while (true) {
-						result = f_readdir_get_filepointer(&staticDIR, &staticFNO,
-						                                   &thisFilePointer); /* Read a directory item */
-						if (result != FR_OK || staticFNO.fname[0] == 0) {
-							break; // Break on error or end of dir
+		bool randomize_all = false;
+		int32_t maxrow = 8;
+		if(Buttons::isShiftButtonPressed())
+		{
+			maxrow = getCurrentInstrumentClip()->noteRows.getNumElements();
+			randomize_all = true;
+		}
+
+
+		String chosen_fn;
+		chosen_fn.set("Nothing to randomize.");
+
+		for (int32_t i = 0; i < maxrow; i++) {
```

Comment:
this loops should probably go from `firstRandomRow` to `lastRandomRow`, where those two variables are set to either `getCurrentInstrumentClip()->getNoteRowOnScreen(lastAuditionedYDisplay, currentSong)->y` (for the single random case, you'll need to check if the returned note row is null as well and bail in that case) or `firstRandomRow=0`/`lastRandomRow=getCurrentInstrumentClip()->noteRows.getNumElements()` for the randomize all case.

Reply:
Single random:
OK on the null check.
But why is: getCurrentInstrumentClip()->getNoteRowOnScreen(lastAuditionedYDisplay, currentSong)->y
better than: getCurrentInstrumentClip()->getNoteRowOnScreen(i, currentSong)

All Random:
I think what you describe is exactly what is happening, but maybe I am overlooking something?

Reply:
> But why is: getCurrentInstrumentClip()->getNoteRowOnScreen(lastAuditionedYDisplay, currentSong)->y
better than: getCurrentInstrumentClip()->getNoteRowOnScreen(i, currentSong)

It avoids running the loop for more iterations than necessary. The idea is to compute the exact range of rows, rather than needlessly iterating over all visible rows. It also allows the `if (randomize_all){}` check in the loop to go away, since both the `randomize_all` and single row cases will treat `i` as the row index.

>   I think what you describe is exactly what is happening, but maybe I am overlooking something?

You're correct, I missed the line that updates `maxRows`.

Reply:
Ah, but

GetCurrentInstrumentClip()->getNoteRowOnScreen(lastAuditionedYDisplay, currentSong)->y

Will only give me the LAST auditioned row - the current code allows (as did the OG drum randomizer) holding several audition pads and randomizing all those rows at once, which is a functionality I would not like to drop.

With the current code

`for (int32_t i = 0; i < maxrow; i++) {
				if (randomize_all || auditionPadIsPressed[i])`

there is very little overhead for the non-pressed pads anyway. I do not see how this could be any more efficient while keeping the multiple press functionality.

Reply:
ah I see, in that case the current logic is fine.